stable_baselines/common/callbacks.py

```diff
@@ -0,0 +1,226 @@
+import os
+from abc import ABC, abstractmethod
+from typing import Union, List
+
+import gym
+import numpy as np
+
+from stable_baselines.common.base_class import BaseRLModel # pytype: disable=pyi-error
+from stable_baselines.common.vec_env import VecEnv, sync_envs_normalization
+from stable_baselines.common.evaluation import evaluate_policy
+
+
+class BaseCallback(ABC):
+    """
+    Base class for callback.
+
+    :param verbose: (int)
+    """
+    def __init__(self, verbose=0):
+        super(BaseCallback, self).__init__()
+        self.model = None
+        self.training_env = None
+        self.n_calls = 0
+        self.num_timesteps = 0
+        self.verbose = verbose
+
+    def init_callback(self, model: BaseRLModel) -> None:
+        """
+        Initialize the callback by saving references to the
+        RL model and the training environment for convenience.
+        """
+        self.model = model
+        self.training_env = model.get_env()
+
+    def on_training_start(self, locals_: dict, globals_: dict) -> None:
+        pass
+
+    # Should we include that?
+    # def on_rollout_start(self, locals_: dict, globals_: dict) -> None:
+    #     pass
+
+    @abstractmethod
+    def on_step(self, locals_: dict, globals_: dict) -> bool:
+        """
+        TODO: Should we modify current implementation?
+        i.e. call after each env step instead after each rollout (current implementation)?
```

Comment:
A per-step callback would be useful (even if it does get called once per step on a VecEnv). Sometimes you might want to kill training at exact number of steps, and algorithms like DQN do not use concept of "rollout" the same way as PPOs/SACs and such.

Reply:
I like the idea, especially since it'd make the interface more consistent between policy-gradient/AC algorithms and Q-learning, but I'm not sure how you'd stop training of a policy-gradient algorithm at arbitrary steps. Normally they're batch-based. I think the TensorFlow code might even demand a fixed batch-size, but even if not the gradient would be really high variance if e.g. you stopped after only one timestep into a batch.

Reply:
> but I'm not sure how you'd stop training of a policy-gradient algorithm at arbitrary steps

If you stop training in the middle of a rollout, then I would discard the rollout for PPO (the training is stopped anyway), otherwise you may end up in a complicated-to-resume situation.